accepted/0000-subscription-matching-in-public-clouds.md

```diff
@@ -0,0 +1,204 @@
+            }
+        }
+    }
+}
+```
+
+All the attributes on the above JSON example are currently required by SUMA to properly create the entries on the DB. Since we don't know the real hardware used to virtualize the instances (node info), those values like `ramMb`, `cpuMhz`, `totalCpuCores`, `totalCpuSockets` have been faked to 0. As mentioned before, with some adaptations in the Java side we could not require those values (similar to the KUBERNETES case).
+
+NOTE: All the "vms" that belongs to the same "Virtual Host Manager", are considered in the same "Virtualization group" for subscription matching. In the case of Public Clouds, this means a customer needs to add a different "Virtual Host Manager" in SUSE Manager for each one of the "zones" of the particular Public Cloud provider where customer wants to gather virtual instances. The "Terms & Conditions" explicitly mentions grouping by "Public Cloud zone" (and not account). Therefore, a customer using i.a. AWS will need to create as many "Virtual Host Manager" on SUSE Manager as "zones" they want to gather. Then each one of those "Virtual Host Manager" is considered as independent Virtualization Group for "1-2 subscriptions" matching.
+
+## Add new Virtual Instance Types to the DB
+New virtual instance types would be needed on the `rhnVirtualInstanceType` database table, these would be `azure`, `aws` and `gce` to indicate the type of the instance.
+
+Also, keep in mind that on Public Clouds we don't really know the architecture of the node that is running the Virtual Instances (we only know the arch of the VM itself) so it might be necessary to add new types of CPU Arch ("rhnCpuArch" DB table) and "Server Arch" ("rhnServerArch") to allow those new types of systems. We could add one type per public cloud provider, i.a. aws, azure, etc, or maybe go with a generic type called `cloud`.
+
+An important thing is also to add the corresponding entry to `rhnServerServerGroupArchCompat` database table to allow `FOREIGN_ENTITLEMENT` to be assigned to any of these new architectures.
+
+## Gather virtual instances and "instance id" from Public Clouds using API
+
+Neither EC2, GCE or Azure expose the smbios "uuid" from the virtual instance through the public API, but instead they identify each instance using an "instance id" which is unique on that particular Public Cloud.
+
+Examples:
+
+- AWS EC2: i-1234567890abcdef0
+- GCE: 152986662232938449
+- Azure: 13f56399-bd52-4150-9748-7190aae1ff21
+
+Since the `virtual-host-gatherer` would only deal with the public APIs, each virtual instance will be identified by its "instance id" on the JSON output from the `virtual-host-gatherer` execution. In order to properly match this "instance id" with the "uuid" of a registered system, this will need some ajustments as describer later on this RFC.
+
+### Azure module:
+
+This is an example about how to deal with Azure Public API to collect the virtual instances, using Azure Python SDK:
```

Comment:
Should we use each cloud's own API, or rather use something like cloud-init, which can provide information for a lot of clouds already?
https://cloudinit.readthedocs.io/en/latest/topics/instancedata.html
https://cloudinit.readthedocs.io/en/latest/topics/datasources.html#datasource-documentati

Reply:
AFAIU, `cloud-init` is a tool that runs inside the virtual instance basically for provisioning the cloud instance at startup time. It seems it gathers common virtual instance information from the "metadata" service (internal API) like the custom grain example I made.

But, we need some tool on the SUSE Manager server side that communicates with the Public Cloud provider using credentials and the external API to gather the instances (registered or not), and, at a first glance, it seems `cloud-init` does not do that.

Reply:
> AFAIU, `cloud-init` is a tool that runs inside the virtual instance basically for provisioning the cloud instance at startup time. It seems it gathers common virtual instance information from the "metadata" service (internal API) like the custom grain example I made.
> 
> But, we need some tool on the SUSE Manager server side that communicates with the Public Cloud provider using credentials and the external API to gather the instances (registered or not), and, at a first glance, it seems `cloud-init` does not do that.

Darn, you are right. I saw the cloud-init docs with exactly the information we need and was too happy. And I cannot find how to get the instance id in libcloud:
https://libcloud.apache.org/

:-(

Reply:
Actually, we could use `libcloud` to deal with AWS, Azure and Google Compute. The `list_nodes` method should return the list of instances including the `id` of the instance:

- AWS: https://libcloud.readthedocs.io/en/latest/compute/drivers/ec2.html
- Azure: https://libcloud.readthedocs.io/en/latest/compute/drivers/azure.html
- Google Compute: https://libcloud.readthedocs.io/en/latest/compute/drivers/gce.html

This way, we would only require `python3-apache-libcloud` as required dependency for dealing all the Public Clouds. I've created an example here: https://gist.github.com/meaksh/7153eba4ae6d98c9c6133e6790c7542d

If you're ok, I would change the RFC examples to make use of `libcloud`

Reply:
I'm OK with whatever makes our life easier and gives the customer the best experience :-)

Will the use of libcloud give us magical support for the 50+ clouds libcloud supports, or will that require additional work on our side? How much work?
https://libcloud.readthedocs.io/en/latest/supported_providers.html